server/src/main/java/io/crate/planner/optimizer/rule/DeduplicateOrder.java

```diff
@@ -68,7 +70,8 @@ public LogicalPlan apply(Order plan,
                              Captures captures,
                              TableStats tableStats,
                              TransactionContext txnCtx,
-                             NodeContext nodeCtx) {
+                             NodeContext nodeCtx,
+                             Function<LogicalPlan, LogicalPlan> resolvePlan) {
         Order childOrder = captures.get(this.childOrder);
         return plan.replaceSources(childOrder.sources());
```

Comment:
Shouldn't this also use `resolvePlan` on the sources ?

Reply:
The `GroupReference` contains the original outputs of the referenced Plan. Only if we need something else e.g. `numExpectedRows` or the relation names but the outputs we need to resolve them. I guess this is not very clear from an outside perspective.

Reply:
This deserves some proper documentation.

Reply:
And also add more exceptions on the unsupported operations.

Reply:
I'd prefer if this possible to somehow force the use of the `resolvePlan` everywhere, and just pass identify func when you don't need to apply something.

Reply:
The problem is that `resolvePlan` is an expensive operation, so we want to minimize its use.

Reply:
That's way I said to pass identity function when not needed, but don't know how feasible that is.

Reply:
We do that already.

Reply:
I think it's fine to require the rule to know when to use it, but the parameter documentation needs to be clear.
Based on the new info I'd suggest something like:

```
    /**
     * @param resolveLogicalPlan Must be used on a source of {@code plan} if the rule needs to access any properties of the source other than the outputs. This may materialize the sub-tree for the source and could be expensive. If the rule doesn't access source properties, don't call it.
     **/
```